### PR TITLE
fixes parameter resolution and server deserializing in case of a null or empty server

### PR DESCRIPTION
**Pull request by gracekarina**
Fixes parameter resolution and server deserializing in case of a null or empty server.

### CODE REVIEW

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java [hunk 1]
```diff
@@ -319,6 +320,7 @@ public void pathsResolver(@Injectable final List<AuthorizationValue> auths) thro
         Assert.assertNotNull(result);
         final OpenAPI openAPI = result.getOpenAPI();
         Assert.assertNotNull(openAPI);
+        //System.out.println("Before Resolver: " + openAPI.getPaths().get("/store/inventory").getGet().getParameters());
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: can you clean these up?

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java [hunk 2]
```diff
@@ -334,7 +336,7 @@ public void pathsResolver(@Injectable final List<AuthorizationValue> auths) thro
         Assert.assertNull(openAPI.getPaths().get("/pet").getParameters());
 
         //remote ref pathItem
-        Assert.assertNull(openAPI.getPaths().get("/pathItemRef").getParameters());
+        //Assert.assertNull(openAPI.getPaths().get("/pathItemRef").getParameters());
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: ... here too

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java [hunk 3]
```diff
@@ -349,6 +351,10 @@ public void pathsResolver(@Injectable final List<AuthorizationValue> auths) thro
         //internal pathItem -> operation -> requestBody
         Schema id = (Schema) openAPI.getPaths().get("/pet/findByStatus").getGet().getRequestBody().getContent().get("multipart/mixed").getSchema().getProperties().get("id");
         assertEquals(id.get$ref(),"#/components/schemas/Pet");
+
+        //internal parameter url
+        //System.out.println("After Resolver: " + openAPI.getPaths().get("/store/inventory").getGet().getParameters());
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: ... and here

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/v3/util/OpenAPIDeserializerTest.java [hunk 4]
```diff
@@ -216,6 +216,38 @@ public void readServerObject(JsonNode rootNode) throws Exception {
...
+
+        final OpenAPIDeserializer deserializer = new OpenAPIDeserializer();
+        final SwaggerParseResult result = deserializer.deserialize(rootNode);
+
+        Assert.assertNotNull(result);
+
+        final OpenAPI openAPI = result.getOpenAPI();
+        Assert.assertNotNull(openAPI);
+
+        assertEquals(openAPI.getServers().get(0).getUrl(),"/");
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/v3/util/OpenAPIDeserializerTest.java]: I would double-check this behavior with @webron.  I'm not sure if we should assume anything if there is no value.

File: modules/swagger-parser-v3/src/test/resources/oas2.yaml [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+      properties:
+        id:
+          type: integer
+          format: int64
+        category:
+          "$ref": "#/components/schemas/Category"
+        user:
+          type: string
+          example: doggie
+          "$ref": "http://localhost:${dynamicPort}/remote/schema#/components/schemas/User"
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/resources/oas2.yaml]: Same with previous comment about `{dynamicPort}` in this document

File: modules/swagger-parser-v3/src/test/resources/oas3.yaml [hunk 6]
```diff
@@ -66,7 +66,7 @@ tags:
     url: http://swagger.io
 paths:
   "/pathItemRef":
-    "$ref": 'http://localhost:${dynamicPort}/remote/path'
+    "$ref": 'http://localhost:${dynamicPort}/remote/path#/paths/~1pet'
```
Comment (fehguy) [inline modules/swagger-parser-v3/src/test/resources/oas3.yaml]: Same as previous comment

File: modules/swagger-parser-v3/src/test/resources/oas2.yaml (hunk 5 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/resources/oas2.yaml]: What do you mean about DynamicPort exactly?

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/v3/util/OpenAPIDeserializerTest.java (hunk 4 above)
Comment (webron) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/v3/util/OpenAPIDeserializerTest.java]: This seems right, based on https://github.com/OAI/OpenAPI-Specification/blob/3.0.0-rc2/versions/3.0.md#oasServers.

File: modules/swagger-parser-v3/src/test/resources/oas3.yaml (hunk 6 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/resources/oas3.yaml]: Changed the name of file, and added template

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java (hunk 1 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: Done

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java (hunk 2 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: Done

File: modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java (hunk 3 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/java/io/swagger/parser/test/OpenAPIResolverTest.java]: Done

File: modules/swagger-parser-v3/src/test/resources/oas2.yaml (hunk 5 above)
Comment (gracekarina) [inline modules/swagger-parser-v3/src/test/resources/oas2.yaml]: Changed the name of file, and added template

**Comment by fehguy** [comment]
@gracekarina I think this is great.  Merging...

**Pull request merged**